File: 03rd/solver.js
const path = require('path');
const fs = require('fs');

function readFile(file) {
    const filePath = path.join(__dirname, file);
    return fs.readFileSync(filePath, 'utf8');
}

function uncorupt(data) {
    return data.match(/mul\(\d+,\d+\)/g);
}

function executeInstructions(instructionSets) {
    sum = 0;
    for (let i = 0; i < correct.length; i++) {
        values = correct[i].match(/\d+/g);
        sum += parseInt(values[0]) * parseInt(values[1]);
    }
    return sum;
}

function splitInDoAndDont(data) {
    // For correctly using Lookahead, thx Copilot
    return data.match(/(do(n't)?\(\))?.*?(?=((do(n't)?\(\))|$))/g);
}

function solve1(data) {
    correct = uncorupt(data);
    if (correct === null) {
        return 0;
    }
    return executeInstructions(correct);
}

function solve2(data) {
    let sum = 0;
    const instructionSets = splitInDoAndDont(data);
    sum = solve1(instructionSets[0]);
    if (instructionSets.length > 1) {
        for (let i = 1; i < instructionSets.length; i++) {
            if (instructionSets[i].startsWith("do()")) {
                sum += solve1(instructionSets[i]);
            }
        }
    }
    return sum;
}


//const data = readFile('test_1.txt');
//const data = readFile('test_2.txt');
const data = readFile('input.txt');

console.log("Advent of Code: 03rd");
console.log("Part 1: ", solve1(data));
console.log("Part 2: ", solve2(data));

File: 23rd/solver.py
import os
import networkx as nx

# pylint: disable=missing-function-docstring

def read_file(file):
    file_path = os.path.join(os.path.dirname(__file__), file)
    with open(file_path, 'r', encoding='utf-8') as f:
        return f.read()

def parse_data(data):
    return [line.split("-") for line in data.split("\n")]

def generate_graph(data):
    graph = nx.Graph()
    for line in data:
        node1, node2 = line
        graph.add_edge(node1, node2)
    return graph

def find_cliques_of_size(graph, size):
    # assumes nx finds cliques in descending order of size
    cliques = []
    for clique in nx.enumerate_all_cliques(graph): # thx ai for this function name
        if len(clique) == size:
            cliques.append(clique)
        elif len(clique) > size:
            break
    return cliques

def solve1(graph):
    # find cliques of size 3
    cliques_of_size_3 = find_cliques_of_size(graph, 3)
    adder = 0
    for clique in cliques_of_size_3:
        if clique[0].startswith("t") or clique[1].startswith("t") or clique[2].startswith("t"):
            adder += 1
    return adder

def solve2(graph):
    cliques = nx.find_cliques(graph)
    maximum = 0
    clique = None
    for c in cliques:
        if len(c) > maximum:
            maximum = len(c)
            clique = c
    sorted_clique = sorted(clique)
    return ",".join(sorted_clique)

DATA = generate_graph(parse_data(read_file("input.txt")))
#DATA = generate_graph(parse_data(read_file("test.txt")))

print("Advent of Code: 19th")
print("Part 1:", solve1(DATA))
print("Part 2:", solve2(DATA))
